lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3307,27 +3282,36 @@ class VJPEmitter final
           /*differentiationOrder*/ 1, functionSource);
     }
 
-    // Check and diagnose non-differentiable arguments.
-    for (unsigned paramIndex : range(originalFnTy->getNumParameters())) {
-      if (indices.isWrtParameter(paramIndex) &&
-              !originalFnTy->getParameters()[paramIndex]
-              .getSILStorageType()
-              .isDifferentiable(getModule())) {
-        context.emitNondifferentiabilityError(
-            original, invoker, diag::autodiff_nondifferentiable_argument);
-        errorOccurred = true;
-        return;
-      }
-    }
-    // Check and diagnose non-differentiable results.
-    if (!originalFnTy->getResults()[indices.source]
-            .getSILStorageType()
-            .isDifferentiable(getModule())) {
-      context.emitNondifferentiabilityError(
-          original, invoker, diag::autodiff_nondifferentiable_result);
-      errorOccurred = true;
+    // Check and diagnose non-differentiable original function type.
+    auto diagnoseNonDifferentiableOriginalFunctionType =
```

Comment:
Note: this logic should be ported to `JVPEmitter::visitApply` in https://github.com/apple/swift/pull/26057.
cc @bartchr808

Reply:
Made a note of this to be a TODO item for myself! 😄